server/routes/authRoutes.js

```diff
@@ -1,299 +1,334 @@
+    if (!existinguser) throw Error('user does not exist')
     res.json(existinguser)
   } catch (e) {
-    res.status(400).json({msg: e.message})
+    res.status(400).json({ msg: e.message })
   }
-});
+})
 
 router.post('/addMailing', async (req, res) => {
-  let {email, firstName, lastName} = req.body;
+  let { email, firstName, lastName } = req.body
   try {
-    const existingEmail = await emailList.findOne({email});
+    const existingEmail = await emailList.findOne({ email })
     if (!existingEmail) {
-    await emailList.create({email, firstName, lastName})
+      await emailList.create({ email, firstName, lastName })
     }
-    res.send({ message: 'success'})
+    res.send({ message: 'success' })
   } catch (e) {
-    res.status(400).send({ message: 'please try again'})
+    res.status(400).send({ message: 'please try again' })
   }
-});
+})
 
 router.post('/test', async (req, res) => {
-  let {email, firstName, lastName} = req.body;
+  let { email, firstName, lastName } = req.body
   try {
-    await emailList.create({email, firstName, lastName})
+    await emailList.create({ email, firstName, lastName })
     console.log(keys.mongoURI)
-    res.send({ message: keys.mongoURI})
+    res.send({ message: keys.mongoURI })
   } catch (e) {
-    res.status(400).send({ message: 'please try again'})
+    res.status(400).send({ message: 'please try again' })
   }
-});
+})
 
-  
+router.get('/github', async (req, res) => {
+  // 1) use the code to get token from github
+  const { code } = req.query
+  let githubToken
+  {
+    const body = {
+      client_id: process.env.NEXT_PUBLIC_GITHUB_CLIENT_ID,
+      client_secret: process.env.GITHUB_CLIENT_SECRET,
+      redirect_uri: process.env.GITHUB_REDIRECT_URL,
+      code
+    }
+    const headers = {
+      Accept: 'application/json'
+    }
+    console.log(body)
+    const res = await API.POST('https://github.com/login/oauth/access_token', headers, body)
+    githubToken = res?.access_token
+  }
+  // 2) use the github token to get user details
+  {
+    const githubUser = await API.GET('https://api.github.com/user', {
+      Authorization: `Bearer ${githubToken}`
+    })
+    if (githubUser.email === null) {
+      const emails = await API.GET('https://api.github.com/user/emails', {
+        Accept: 'application/vnd.github+json',
+        Authorization: `Bearer ${githubToken}`,
+        'X-GitHub-Api-Version': '2022-11-28'
+      })
+      let userPrimaryEmail
+      // check if we have an user record with any of the emails , if we do then the user is not new and he is just signing in, if not then the user is new and we need to save his details in db. In this flow we need to update the user record with the new primary email if the user has changed his primary email in github
+      userPrimaryEmail = emails?.find(email => email.primary === true)?.email
+      githubUser.email = userPrimaryEmail
+      githubUser.emails = emails.map(x => x.email)
```

Comment:
we're calling emails here, but if it is undefined, then this will cause error. Can we add a check for if email is undefined?